sync-core/src/main/java/com/cloudant/sync/query/QueryValidator.java

```diff
@@ -437,6 +459,48 @@ private static boolean validateClause(Map<String, Object> clause) {
         return false;
     }
 
+    /**
+     * This method handles the special case where a text search clause is encountered.
+     * This case is special because a $text operator expects a Map value whose key can
+     * only be the $search operator.
+     *
+     * @param clause The text clause to validate
+     * @param textClauseLimitReached A flag used to track the text clause limit
+     *                               throughout the validation process.  The
+     *                               current limit is one text clause per query.
+     * @return true/false whether the clause is valid
+     */
+    @SuppressWarnings("unchecked")
+    private static boolean validateTextClause(Object clause,
+                                              Boolean[] textClauseLimitReached) {
+        Map<String, Object> textClause;
+        if (clause instanceof Map) {
+            textClause = (Map<String, Object>) clause;
+            if (textClause.size() == 1) {
+                String operator = (String) textClause.keySet().toArray()[0];
+                if (operator.equals(SEARCH)) {
+                    Object clauseOperand = textClause.get(operator);
+                    if (textClauseLimitReached[0]) {
+                        String msg = "Multiple text search clauses not allowed in a query.  " +
+                                "Rewrite query to contain at most one text search clause.";
+                        logger.log(Level.SEVERE, msg);
+                    } else {
+                        textClauseLimitReached[0] = true;
+                        return validatePredicateValue(clauseOperand, operator);
+                    }
+                } else {
+                    String msg = String.format("Invalid operator %s in text search", operator);
+                    logger.log(Level.SEVERE, msg);
+                }
+            }
+        } else {
+            String msg = String.format("Text search expects a Map, found %s instead.", clause);
+            logger.log(Level.SEVERE, msg);
+        }
```

Comment:
Can you structure this as a series of tests rather than a nested if/else clause? It's easy to miss that `return validatePredicateValue(...)` in the middle.

```java
if (!(clause instanceof Map)) { 
    String msg = String.format("Text search expects a Map, found %s instead.", clause);
    logger.log(Level.SEVERE, msg);
    return false;
}

/* etc. */
```